### PR TITLE
add config field in MLToolSpec for static parameters

### PR DESCRIPTION
**Pull request by jngz-es**
### Description
Add config field in MLToolSpec to support static parameters in tool execution.

An example from the issue 2918.
```
POST /_plugins/_ml/agents/_register
{
	"name": "Test Agent",
	"type": "conversational",
	"description": "test agent",
	"llm": {
		"model_id": "<model id>",
		"parameters": {
      "max_iteration": 5,
      "stop_when_no_tool_found": true,
      "response_filter": "$.completion"
    }
	},
	"memory": {
		"type": "conversation_index"
	},
	"app_type": "chat",
	"tools": [
		{
			"type": "SearchIndexTool",
			"description": "A tool to search opensearch index with natural language question.",
			"include_output_in_agent_response": true,
			"config": {
				"input": "{\"index\": \"sample-index\", \"query\": {\"query\": { \"match_all\": {}}} }"
			}
		}
	]
}
```

### Related Issues
https://github.com/opensearch-project/ml-commons/issues/2836
https://github.com/opensearch-project/ml-commons/issues/2918

### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [ ] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/ml-commons/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by dbwiddis — COMMENTED** [review]
Can you open an issue on Flow Framework repo amplifying what changed here? We will need to update the [`ToolStep`](https://github.com/opensearch-project/flow-framework/blob/43a9f871de50fc57758efd57c0f8899b22cecc3a/src/main/java/org/opensearch/flowframework/workflow/ToolStep.java#L51-L94) to account for a change in API.  Better yet, open a PR with the changes!

Also please make sure you update the [API specification](https://github.com/opensearch-project/opensearch-api-specification/blob/c6c0c417b19dd9b63c4e37379b7e94d412be7ea8/spec/schemas/ml._common.yaml#L271-L280) which Flow Framework will be using to validate these API parameters.   (It's a checklist item at the top of this PR.)

CC: @owaiskazi19 @amitgalitz @junweid62 @ylwu-amzn

File: common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java:85]: Add version checking for BWC

**Comment by jngz-es** [comment]
The flow framework issue, https://github.com/opensearch-project/flow-framework/issues/878.

File: common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java
Comment (jngz-es) [reply common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java]: Good catch, done.

File: common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java
Comment (dbwiddis) [inline common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java:29]: Isn't 2.17.0 already released so this would be a 2.18.0 feature?

File: common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java
Comment (dbwiddis) [inline common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java:74]: This seems odd here, it looks like we're allowing null keys and null values in the map which I doubt is the case (at least the keys shouldn't be optional, and I'm not sure of the benefit of null values vs. just omitting the key).  Same comment in the `writeTo()` below.

File: common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java
Comment (dbwiddis) [inline common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java:115]:
```suggestion
        if (configMap != null && !configMap.isEmpty()) {
```

File: common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java
Comment (dbwiddis) [inline common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java:152]: may want to rename `getParameterMap` to be more general `getStringToStringMap` or similar.

File: common/src/test/java/org/opensearch/ml/common/agent/MLAgentTest.java
Comment (dbwiddis) [inline common/src/test/java/org/opensearch/ml/common/agent/MLAgentTest.java:49]:
```suggestion
            List.of(new MLToolSpec("test", "test", "test", Collections.EMPTY_MAP, false, Collections.emptyMap())),
```
Use `emptyMap()` rather than `EMPTY_MAP` to get built in type validation.  This will probably create unchecked cast warnings.  (Same comment multiple times below.)

File: common/src/test/java/org/opensearch/ml/common/agent/MLToolSpecTest.java
Comment (dbwiddis) [inline common/src/test/java/org/opensearch/ml/common/agent/MLToolSpecTest.java:77]: I don't think the order of fields in JSON is guaranteed unless you specify a sorting order.  Same comment in other assertions.

File: common/src/test/java/org/opensearch/ml/common/agent/MLToolSpecTest.java
Comment (dbwiddis) [inline common/src/test/java/org/opensearch/ml/common/agent/MLToolSpecTest.java:126]: Nit: using the same string everywhere can let copy-paste-and-forget-to-change errors slip through.  Recommend using a different string for each field.

File: common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java
Comment (jngz-es) [reply common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java:74]: For the key, changed from `readOptionalString/writeOptionalString` to `readString/writeString`.

File: common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java
Comment (jngz-es) [reply common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java:115]: Done.

File: common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java
Comment (jngz-es) [reply common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java:152]: It's an existing function, I prefer to keep it unchanged at this moment.

File: common/src/test/java/org/opensearch/ml/common/agent/MLAgentTest.java
Comment (jngz-es) [reply common/src/test/java/org/opensearch/ml/common/agent/MLAgentTest.java:49]: Makes sense, changed.

File: common/src/test/java/org/opensearch/ml/common/agent/MLToolSpecTest.java
Comment (jngz-es) [reply common/src/test/java/org/opensearch/ml/common/agent/MLToolSpecTest.java:77]: Do you have any suggestions?

File: common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java
Comment (jngz-es) [reply common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java:29]: We want to catch up with 2.17 patch.

Comment (dbwiddis) [reply common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java:29]: 2.17.0 open source is already released.  If you want to change the version on the back port to that branch that gets synced elsewhere that’s a reasonable option but for anyone consuming the maven central artifact BWC would break with that version check.

File: common/src/test/java/org/opensearch/ml/common/agent/MLToolSpecTest.java
Comment (dbwiddis) [reply common/src/test/java/org/opensearch/ml/common/agent/MLToolSpecTest.java:77]: I usually do assertions that the string contains each key value pair.  Or just extract to a map with a util function and check keys.   I think there are some optional arguments with the Xcontent that let you specify an ordering.

Comment (dbwiddis) [reply common/src/test/java/org/opensearch/ml/common/agent/MLToolSpecTest.java:77]: Looked into this.  Seems internal comments indicate the "XContent" preserves the same ordering, so since you're just parsing the same thing you built, this should be fine in this test case.  The concern would be if you're using any sort of external parsing using a map, where the ordering isn't guaranteed.

File: common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java
Comment (dbwiddis) [reply common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java:29]: Actually looks like a 2.17.1 release is happening so it should be changed to that.

Comment (dbwiddis) [reply common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java:29]: Except they don't want any new features, just critical bugfixes. So nevermind :|

**Comment by yuye-aws** [comment]
@reuschling Can you help review this PR?

File: common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java
Comment (yuye-aws) [reply common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java:29]: This PR is not a security fix. We can leave it to 2.18.

**Comment by yuye-aws** [comment]
@jngz-es Can you please provide an example on how to use tool `config`?

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLChatAgentRunner.java
Comment (yuye-aws) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLChatAgentRunner.java:484]: I would suggest extract` parameters.putAll(toolSpecMap.get(action).getConfigMap());` outside of the if-else statement.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLConversationalFlowAgentRunner.java
Comment (yuye-aws) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLConversationalFlowAgentRunner.java:442]: The function is the same as MLFlowAgentRunner. Please extract the function to classes like `AgentUtils`.

File: ml-algorithms/src/test/java/org/opensearch/ml/engine/algorithms/agent/MLFlowAgentRunnerTest.java
Comment (yuye-aws) [inline ml-algorithms/src/test/java/org/opensearch/ml/engine/algorithms/agent/MLFlowAgentRunnerTest.java:320]: Please also verify that non-config parameters can override

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLConversationalFlowAgentRunner.java
Comment (reuschling) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLConversationalFlowAgentRunner.java:442]: As far as I see, this PR writes all parameters from the new config map into the parameter map that is relevant for execution, which means that config parameters will overwrite existing ones with same name, i.e. have prioritization. Thus, specifying an own static input parameter as suggested in #2836 and #2918 is possible, but crucial is also the possibility to specify placeholders inside the new config. I can not see some substitutions here but maybe this takes place somewhere else, for all parameters?

For #2836, the placeholder `"${parameters.MLModelTool.output}"` was asked for.
For #2918, SearchIndexTool inside conversational agent, there is the need of a placeholder for the **last action input** of the agent, that must inserted into the OpenSearch query defined statically inside the new config. If conversational agent just sets the input parameter for the next tool, I doubt this information is lost if I specify a static input inside the tool config. 

Last but not least, the documentation should be adjusted accordingly, e.g. [SearchIndexTool](https://opensearch.org/docs/latest/ml-commons-plugin/agents-tools/tools/search-index-tool/) and [tutorial example](https://opensearch.org/docs/latest/ml-commons-plugin/tutorials/rag-conversational-agent/#exposing-only-the-question-parameter). Also some list or hint to the common available placeholders would be nice, in order to know e.g. how the placeholder for the former conversational agent step action input looks like.

**Comment by jngz-es** [comment]
> @jngz-es Can you please provide an example on how to use tool `config`?

Provided an example in the description section.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLChatAgentRunner.java
Comment (jngz-es) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLChatAgentRunner.java:484]: Don't get it. I need to add the config parameters override immediately before tool run.

File: common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java
Comment (jngz-es) [reply common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java:29]: Based on the discussion with Yuye, I will make the version 2.18. Thanks @yuye-aws

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLConversationalFlowAgentRunner.java
Comment (jngz-es) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLConversationalFlowAgentRunner.java:442]: Good callout, I am thinking of supporting placeholders inside the config. Thanks @reuschling

File: common/src/test/java/org/opensearch/ml/common/agent/MLToolSpecTest.java
Comment (jngz-es) [reply common/src/test/java/org/opensearch/ml/common/agent/MLToolSpecTest.java:126]: done

File: common/src/test/java/org/opensearch/ml/common/agent/MLToolSpecTest.java
Comment (jngz-es) [reply common/src/test/java/org/opensearch/ml/common/agent/MLToolSpecTest.java:25]: done

File: common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java
Comment (Zhangxunmt) [inline common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java:52]: Is it valid that all the configs are Strings? Is it better to use Map<String, Object> here?

**Review by Zhangxunmt — APPROVED** [review]

File: common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java
Comment (dhrubo-os) [reply common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java:52]: +1 on this

File: common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java:74]: why `readOptionalString` ? so a key might not have a value?

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/AgentUtils.java
Comment (dhrubo-os) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/AgentUtils.java:478]: Don't we have any test class for this?

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLFlowAgentRunner.java
Comment (dhrubo-os) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLFlowAgentRunner.java:291]: rather than putting same piece of code, can we try to use a common method `getToolExecuteParams` for both MLFlowAgentRunner & MLConversationalFlowAgentRunner?

**Comment by yuye-aws** [comment]
Pinging @reuschling to review the PR.

File: common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java
Comment (dbwiddis) [reply common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java:52]: -1 on that.  The parameters map is already a `Map<String, String>`.  In the initial PR comment an example is given of a much more complex "object" but escaped as a String.

Trying to parse anything else from Flow Framework will get a lot more complex without typed objects to construct from the JSON.

File: common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java
Comment (dbwiddis) [reply common/src/main/java/org/opensearch/ml/common/agent/MLToolSpec.java:74]: Yes, this allows you to override a value with a non-value.

**Comment by dbwiddis** [comment]
> Pinging @reuschling to review the PR.

Thanks @yuye-aws for prompting review on this. Thanks @Zhangxunmt for approval. Thanks @dhrubo-os for your initial review.

Code freeze / 2.18 RC generation is next Tuesday Oct 22.  Downstream plugin PRs (https://github.com/opensearch-project/flow-framework/pull/899) are dependent on this one being merged and need to merge before code freeze.  

Please do not wait until the last minute to get this merged; last-minute PR merges have occurred in at least 2 previous releases and this one's been open for a few weeks.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLConversationalFlowAgentRunner.java
Comment (reuschling) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLConversationalFlowAgentRunner.java:442]: I'm not sure if we have a chicken-and-egg problem here regarding the 'input' parameter. In #2918 'SearchIndexTool inside conversational agent', there is the need of a placeholder for the last output (i.e. action input) of the former agent step. This output/action input must be inserted into the OpenSearch query defined statically inside the new config, by placeholder.

There exists two possibilities for this:
1. There is an existing placeholder reference to the last output/action input that can be used, i.e. a dedicated parameter list entry. I'm not aware of such an entry, but if such a placeholder exists, the current solution would fix this scenario with a static config of e.g. ["input='"query": {"match": {"body": {"query": "${someUnknownPlaceholderToActionInput}"}}'"].
2. There is no such a placeholder, the last output/action input is just the 'input' entry in the current 'params' method parameter - so the placeholder inside the static config would be 'input'. This value inside 'params' is currently overwritten with 'input' from the static config, **before** substitution. So it is lost, right?

The final behavior should be as follows:
* method params: "input='What is Frodos age?'"  (comes from former agent step)
* static config params "input='"query": {"match": {"body": {"query": "${parameters.input}"}}'"

=> final params for tool execution should be then
"input='"query": {"match": {"body": {"query": "What is Frodos age?"}}'"

This could be achieved as follows:
```java
       // remember former input
       String formerInput = executeParams.get("input");

        if (toolSpec.getConfigMap() != null && !toolSpec.getConfigMap().isEmpty()) {
            executeParams.putAll(toolSpec.getConfigMap());
        }

        if (executeParams.containsKey("input")) {
            String input = executeParams.get("input");

            // set former input as substitution value
           if(formerInput != null)
               executeParams.put("input", formerInput); // add it as "formerInput" or "actionInput" to introduce a new placeholder

            StringSubstitutor substitutor = new StringSubstitutor(executeParams, "${parameters.", "}");
            input = substitutor.replace(input);
            executeParams.put("input", input);
```

The lack is that you can not reference to the input specified in the new static config anymore, which is maybe wanted behavior. To overcome this, we could introduce a new placeholder for the former input (aka the placeholder I mentioned in 1.). This would be a new parameter list entry, e.g. 'formerInput' (just add the formerInput with `executeParams.put("formerInput", formerInput);`).

The static config could be defined now as ["input='"query": {"match": {"body": {"query": "${parameters.formerInput}"}}'"].

Another name for 'formerInput' could be 'actionInput', maybe this is more intuitive.


The same is true for `MLFlowAgentRunner` of course, and for `AgentUtils`, whereby in AgentUtils the former input is defined in the method parameter 'actionInput'. So the only change here would be `executeParams.put("formerInput", actionInput);` or `executeParams.put("actionInput", actionInput);` respectively.

**Review by reuschling — CHANGES_REQUESTED** [review]

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLConversationalFlowAgentRunner.java
Comment (jngz-es) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLConversationalFlowAgentRunner.java:442]: Thanks @reuschling for the detailed comment. Actually we are the first possibility you mentioned.
#### Flow Agent
We have a dedicated parameter parameter for the previous tool output here https://github.com/opensearch-project/ml-commons/blob/main/ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLFlowAgentRunner.java#L112. So you can reference it in the current tool input parameter as ${parameters.previous_tool_name.output}.
#### Conversational Agent
We take the action input for tools only from LLM response, i.e, the agent totally depends on LLM to do the ReAct. User can define a tool input to override the action input, but I don't see a use case for a placeholder of action input.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/AgentUtils.java
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/AgentUtils.java:482]: Add these
```
            if (isJson(actionInput)) {
                Map<String, String> params = getParameterMap(gson.fromJson(actionInput, Map.class));
                toolParams.putAll(params);
            }
```

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLConversationalFlowAgentRunner.java
Comment (reuschling) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLConversationalFlowAgentRunner.java:442]: Good to hear that there is already a parameter for flow agent.
For Conversational agent, my example is exactly my use case. I want to predefine an Opensearch query, and the tool description says what's to insert. Thus, the previous llm step (the agent) doesn't have to generate the opensearch syntax, only a simple drop in, so I can guarantee this query will be executed, under all circumstances.
More specifically, I want to e. g. count documents with a query string inside.
Without a placeholder, the agent would have no possibility to adjust the query - only static queries independent from the previous steps in the thought chain would be possible.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/AgentUtils.java
Comment (jngz-es) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/AgentUtils.java:482]: good catch, added

**Review by ylwu-amzn — APPROVED** [review]

**Review by ylwu-amzn — APPROVED** [review]

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLConversationalFlowAgentRunner.java
Comment (reuschling) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLConversationalFlowAgentRunner.java:442]: @jngz-es any comment to my use case? This is my origin need and the reason behind #2918

Comment (jngz-es) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLConversationalFlowAgentRunner.java:442]: Got your point. I created another issue to track it. https://github.com/opensearch-project/ml-commons/issues/3134

Comment (reuschling) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLConversationalFlowAgentRunner.java:442]: Is there a chance left that it make it to 2.18?

Comment (yuye-aws) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLConversationalFlowAgentRunner.java:442]: @reuschling The backport PR has been merged into the 2.x branch. The 2.18 branch will include this PR.

Comment (reuschling) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLConversationalFlowAgentRunner.java:442]: I mean the conversational agent placeholder now tracked in #3134

Comment (jngz-es) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLConversationalFlowAgentRunner.java:442]: Hi @reuschling , I raised a pr https://github.com/opensearch-project/ml-commons/pull/3200, could you take a look?

**Pull request merged**
